### ISSUE TITLE
Problem during importing relations with auto indexing

### ISSUE THREAD

**Issue by mkleszcz** [jexp/batch-import#37]
Hi,
I'm trying to import nodes and I've got following error:

```
    Importing 20000 Nodes took 3 seconds

    Importing 118 Nodes took 0 seconds

    Total import time: 9 seconds
    Exception in thread "main" java.util.NoSuchElementException: More than one element in org.mapdb.Bind$5$1@1a0225b. First element is '1' and the second element is '2'
    at org.neo4j.helpers.collection.IteratorUtil.singleOrNull(IteratorUtil.java:122)
    at org.neo4j.helpers.collection.IteratorUtil.singleOrNull(IteratorUtil.java:289)
    at org.neo4j.batchimport.index.LongIterableIndexHits.getSingle(LongIterableIndexHits.java:33)
    at org.neo4j.batchimport.index.LongIterableIndexHits.getSingle(LongIterableIndexHits.java:12)
    at org.neo4j.batchimport.Importer.lookup(Importer.java:100)
    at org.neo4j.batchimport.Importer.id(Importer.java:139)
    at org.neo4j.batchimport.Importer.importRelationships(Importer.java:115)
    at org.neo4j.batchimport.Importer.doImport(Importer.java:188)
    at org.neo4j.batchimport.Importer.main(Importer.java:74)
```

I've got in batch.properties:
        batch_import.node_index.nodetype=exact
        batch_import.node_index.activity_id=exact
        batch_import.node_index.student_id=exact

I'm importing nodes from two files. students.csv:

```
    nodetype:string:nodetype    id_student:int:student_id   student_name
    student 62296   cc65ccc2d7a1a634d53
    student 62297   32731e2b3f905201751
    student 62298   f467e881fb927293bfd
    student 62299   f58a8028ca573feb8d5
    ....
```

and main.csv:

```
    nodetype:string:nodetype    [...]   id:int:activity_id
    act [...]   1
    act [...]   2
    act [...]   3
    act [...]   4
    act [...]   5
    act [...]   6
    act [...]   7
```

my relations file:

```
    id:int:activity_id  id_student:int:student_id   type
    1   62296   SOLVED_BY
    2   62296   SOLVED_BY
    3   62296   SOLVED_BY
    4   62296   SOLVED_BY
    5   62296   SOLVED_BY
    6   62296   SOLVED_BY
    7   62296   SOLVED_BY
    8   62296   SOLVED_BY
    9   62296   SOLVED_BY
    10  62296   SOLVED_BY
```

Any advice what I'm doing wrong?;)

**Comment by jexp** [comment]
Are you sure you have no duplicates? Seems that the activities might be duplicates per student? The lookup by id for the relationships only work for unique values (actually we could make them work for non-unique values by creating relationships for the cross product - would that be really what you wanted?). 

So you might be needed to introduce a synthentic property for the activity to be indexed uniquely right now.

Are the snippets you shared enough to create the issue?

Please check the activity uniqueness, if they are unique, please try to create a dataset that fails and share it with me.

Thanks

**Comment by mkleszcz** [comment]
I checked the uniqueness. I made a test by placing only one activity and one student with one relation between and it fails the same way.

Files:
https://dl.dropboxusercontent.com/u/11823021/test/activities.csv
https://dl.dropboxusercontent.com/u/11823021/test/students.csv
https://dl.dropboxusercontent.com/u/11823021/test/rels.csv

**Comment by mkleszcz** [comment]
I try to debug code... There are duplicates but in the database. For some reason each node is twice inserted.

**Comment by himmatrana** [comment]
duplicate node inserted

**Comment by jexp** [comment]
Thanks for reporting, I look into it.

**Comment by himmatrana** [comment]
thanks.
last node inserted twice

**Comment by Tooa** [comment]
Same Problem here:

```
Exception in thread "main" java.util.NoSuchElementException: More than one element in org.mapdb.Bind$5$1@1717d968. First element is '26802' and the second element is '27151'
at org.neo4j.helpers.collection.IteratorUtil.singleOrNull(IteratorUtil.java:122)
at org.neo4j.helpers.collection.IteratorUtil.singleOrNull(IteratorUtil.java:289)
at org.neo4j.batchimport.index.LongIterableIndexHits.getSingle(LongIterableIndexHits.java:33)
at org.neo4j.batchimport.index.LongIterableIndexHits.getSingle(LongIterableIndexHits.java:12)
at org.neo4j.batchimport.Importer.lookup(Importer.java:100)
at org.neo4j.batchimport.Importer.id(Importer.java:154)
at org.neo4j.batchimport.Importer.importRelationships(Importer.java:119)
at org.neo4j.batchimport.Importer.doImport(Importer.java:203)
at org.neo4j.batchimport.Importer.main(Importer.java:74)
```

But I do have duplicates for example:

```
identifier:string:TokenNode identifier:string:TokenNode occurence
a    b    OCCURENCE
a    a    OCCURENCE
a    b    OCCURENCE
```

According to the tip above, I tried to add a synthetic property to the relationthip called "id" and initialized with uuid4. Results in following like:

```
identifier:string:TokenNode identifier:string:TokenNode occurence id
a     a    OCCURENCE    515cfbe7-53ab-46a8-9eaa-acaa0d99ce49
```

But still the same Error. What am I doing wrong? Regards.

**Comment by jexp** [comment]
It is about duplicate entries in your nodes-file.

Make sure to pull the latest version too.

**Comment by Tooa** [comment]
I cloned and compiled the repository two days ago. I'm sure my node file hasn't any duplicates. I created it with a map reduce on the count of words. You can see my files here:

https://www.dropbox.com/s/rv2lqgcfti0361a/nodes.csv
https://www.dropbox.com/s/cu2bc37w66xgs2b/relation.csv

Regards. 

Update:

Figured the problem out. "spd\" is imported as "spd" and so a duplicate entry in the node file exists. Don't know if it is a expected behaviour of the batch-importer. Thanks for help. 

Update2:

Same problem. "\o" is equal to "o"

**Comment by szabogtamas** [comment]
I experience the same issue with the 2.0 importer.
I have no duplicates in my csv files, but during creation of relationships, I get the error regarding duplicate nodes.
Was this fixed for the master branch only? Or it was also fixed for the 2.0 branch and I'm doing something else wrong?
Thank you very much.

**Comment by jexp** [comment]
Sorry, Neo4j 2.0 got rid of the Reference Node -> Node 0.

So your node-id's start now with 0, can you adapt your relationship-files and reimport?

Thanks.
